schemas/cloud.sql

```diff
@@ -72,18 +79,103 @@ CREATE PROCEDURE ReplaceCloudRecord(
   disk INT, benchmarkType VARCHAR(50), benchmark DECIMAL(10,3), storageRecordId VARCHAR(255),
   imageId VARCHAR(255), cloudType VARCHAR(255),
   publisherDN VARCHAR(255))
+
 BEGIN
-    REPLACE INTO CloudRecords(VMUUID, SiteID, CloudComputeServiceID, MachineName, LocalUserId, LocalGroupId,
-        GlobalUserNameID, FQAN, VOID, VOGroupID, VORoleID, Status, StartTime, EndTime, SuspendDuration,
-        WallDuration, CpuDuration, CpuCount, NetworkType, NetworkInbound, NetworkOutbound, PublicIPCount, Memory, Disk, BenchmarkType, Benchmark, 
-        StorageRecordId, ImageId, CloudType, PublisherDNID)
+    DECLARE measurementTimeCalculated DATETIME;
+    DECLARE recordCreateTimeNotNull DATETIME;
+
+    IF(status='completed') THEN
+        -- in this case, the recordCreateTime and measurementTime could
+        -- be wildly different as the VM has ended.
+
+        -- if we werent supplied a record create time
+        -- for a completed VM we have decided to use the end time
+        SET recordCreateTimeNotNull = IF(recordCreateTime IS NULL or recordCreateTime='0000-00-00 00:00:00', endTime, recordCreateTime);
+
+        -- Use the end time as the measurement time
+        SET measurementTimeCalculated = endTime;
+        
+    ELSE
+        -- In the case of a running VM, the measurement time will
+        -- equal the record create time
+        IF(recordCreateTime IS NULL or recordCreateTime='0000-00-00 00:00:00') THEN
+            -- Calculate the time of measurement so we can use it later to determine which
+            -- accounting period this incoming record belongs too.
+            SET measurementTimeCalculated = TIMESTAMPADD(SECOND, (IFNULL(suspendDuration, 0) + IFNULL(wallDuration, 0)), startTime);
+            -- We recieve and currently accept messages without a start time
+            -- which causes the mesaurementTimeCalculated to be NULL
+            -- which causes a loader reject on a previously accepted message
+            -- so for now, set it to the zero time stamp as is what happens currently
+            SET measurementTimeCalculated = IFNULL(measurementTimeCalculated, '00-00-00 00:00:00');
+            SET recordCreateTimeNotNull = measurementTimeCalculated;
+        ELSE
+             -- Use the supplied record create time as the measurement time
+            SET measurementTimeCalculated = recordCreateTime;
+            SET recordCreateTimeNotNull = recordCreateTime;
+        END IF;
+    END IF;
+
+    INSERT INTO CloudRecords(RecordCreateTime, VMUUID, SiteID, CloudComputeServiceID, MachineName,
```

Comment:
you can't use `ON DUPLICATE KEY UPDATE` with a `REPLACE`.

So using `REPLACE` would mean the record incoming record always takes precedence over the existing one. But by using `ON DUPLICATE KEY UPDATE` you can tell it to use the one with the highest `MeasurementTime`